### ISSUE TITLE
[IBP-1290] Bulk invoice payments creation API

### ISSUE THREAD

**Issue by bohte** [ledger123/runmyaccounts#566]
Hello,

As we discussed on different channel - we would like to request new API endpoint creation on SQL-Ledger side that will allow to execute bulk invoice payments creation instead of using functionality provided on AR/AP invoice pages:
![image](https://github.com/user-attachments/assets/474907ac-60cb-4e34-ba07-034d79974e9c)

1) To cover all our needs, introduced route must accept and process following request body:
_Example written in JSON (may be also sent as form data which is, i believe, more usual approach)_
`[
   {
      "invoiceId": {TARGET INVOICE ID},
      "datePaid": {DATE VALUE} ('date' field from invoice payment form),
      "source": {STRING VALUE} ('source' field from invoice payment form),
      "memo": {STRING VALUE} ('memo' field from invoice payment form),
      "amount": {NUMBER VALUE} ('amount' field from invoice payment form),
      "exchangeRate": {NUMBER VALUE) ('exch' field from invoice payment form),
      "account": {CHART REFERENCE} ('account' field from invoice payment form)
   },
   {
   ...
   }
]`
2) Type of the invoice: AR, AP, may also be needed. Would be good to make it specified as URL path or query parameter, but it can be also done as JSON/form data payload parameter.
3) As far as I know - login of the user is also required. It can be sent as form data parameter in case of according approach, if we go with JSON - it would better be request header
4) **IMPORTANT**: bulk processing must be transactional, which means that either all requested payments are created or none of them are.

Please, name new branch as:
**IBP-1290-E-Banking-Sub-ledger-Splitting-and-Compatibility-Multiple-invoice-linking-bulk-API**

Thanks in advance!

**Comment by ledger123** [comment]
Hi @bohte ,

I have figured out after testing that I need following information from your application to post the payment through existing SQL-Ledger SL::CP.pm code. You can send multiple array elements for multiple invoices. Is this ok for you?

Field names can be different as I will map them again as needed by SQL-Ledger API. 

`[
  {
    "formname": "receipt",
    "AR": "0111--VAT testing",
    "AR_paid": "1200--Bank Account GBP",
    "arap": "ar",
    "vc": "customer",
    "payment": "payment",
    "type": "receipt",
    "rowcount": 1,
    "checked_1": "1",
    "currency": "GBP",
    "exchangerate": "1",
    "customer_id": "10119",
    "ARAP": "AR",
    "datepaid": "07-10-2024",
    "source": "source123",
    "memo": "memo123",
    "amount": "888",
    "id_1": "10227",
    "transdate_1": "10-11-2007",
    "discount_1": "0",
    "paid_1": "888"
  }
]
`
Please confirm and then I will convert this into a POST end point for you.

**Comment by ledger123** [comment]
Posting this as image too for a bit more clarity:

![image](https://github.com/user-attachments/assets/f5151805-dbe7-4abe-acb8-d5a61e86f5c3)

**Comment by bohte** [comment]
Hi,

I confirm the approach. Please, proceed with implementation.
After first closure is done - we can have a look in direction of improvements.

Thanks!

**Comment by bohte** [comment]
`[  
  {
              "ARAP": "AR", // "AR" or "AP"
    "id": 10000, // Identifier from ar/ap table
    "payment": {
      "date": 01.01.1970, // payment row date
      "amount": 100.00, // payment row amount (paid)
      "source": "Text", // payment row source
      "memo": "Text", // payment row memo
      "exchangeRate": 1.02, // payment row exchange rate
      "account": "1001--Chart of acc" // payment row account
    }
        },
  ...
]`

**Comment by bohte** [comment]
Initial testing is done. Feedback is as follows:

1) Response body adjustments:
- Some payload field names must be changed
- Date format needs to be changed to 'yyyy-MM-dd'
**Status**: I've updated code by myself, here is the [PR](https://github.com/ledger123/runmyaccounts/pull/578) - ✅ 
(please, feel free to merge PR if You confirm changes, otherwise - let me know)

2) Type must be changed for some payload fields:
"id": string -> number
"exchangeRate": string -> number
**Status**: TODO - ❌

3) Multiple payments posting doesn't work (even though response is 200 OK)
**Status**: TODO - ❌

4) Client DB selection is hardcoded, has to be sent as part of request body/URL
**Status**: TODO - ❌

5) Response body is required:
It has to be request body with "payment" object (for each invoice) being enriched with value generated in new column
_Note_: I will introduce script for new column creation by myself and let You know when it is done
**Status**: TODO - ❌

**Comment by bohte** [comment]
Hi Armaghan,
As discussed i'm putting under the ticket all the scenarios that we currently want to achive with important details.
This description is meant to explain the while situation and goals we want to achieve.

_NOTE_: Please, be aware that when I mention importing of bank transactions - I refer to our own funcitonality and it has nothing to do with SQL-Ledger logic. Yet, we do use SQL-Ledger client databases to store imported transactions.


**Description**:
The general idea of the feature we are working on is to allow user to:
- create bookings (book) imported bank transactions
- create invoice payments based on imported bank transactions (aka link imported transaction to invoices)


**Current state**:
At the moment we have already implemented first of bullet-points above, which capability to book imported bank transactions.
Unlike the situation with invoice payments we had no problems there since:

1) After booking is created from imported transaction - the whole booking does represent linked transaction and there are no problems of identifying this booking by simply using "gl" table 'id' column to put any logic on top

(e. g. blocking editability from our UI/API, deleting booking if user decides to unbook imported transaction)
2) Only one GL booking can be created for single imported transaction so it is one-to-one relation


**Where issues begin**:
When it comes to creating payments for existing invoices based on imported transactions, process is much more complicated:

1) One bank transaction can be linked to several invoices, which will mean that based on one imported transaction we can create multiple payments for different invoices
_NOTE_: It was the initial reason for us to request bulk API endpoint, due to posting several invoice payments separately hits performance and responsivness.

2) In terms of data, imported transaction is not only being linked to invoice, but also to certain payment under this particular invoice (technically speaking, inside "imported_transaction_to_booking" table, for invoice payment scenario, we would also need to fill 'booking_part_id' column in order to determine which payment under invoice is based on imported transaction)
_NOTE_: This is the reason why we needed to introduce 'determinant' column, and why we needed to have it's value in the response - to then put it in relation table as 'booking_part_id'.

3) Considering specificity of SQL-Ledger logic of handling payments - inconsistency might appear each and every time invoice is getting updated, which will break data relations cause discrepancy.


**To sum it up**:
We have to find solutions for following issues:
1) Bulk invoice payment posting
2) Consistency maintenance for relations between imported transactions and invoice payments
3) Most likely we will also need to have bulk invoice payment deleting API, which will work in the same manner as posting, but instead of creating sent payments - it will delete invoice payments by identifiers

**Comment by ledger123** [comment]
Hi @bohte ,

I have updated the api.pl with import_transaction_id and also updated sql-ledger code to show the id in payment lines during our testing. Later on we will make it hidden field. Here is screen shot from my server.

I have also updated the curl request for your reference in api.pl which I used for testing.

![image](https://github.com/user-attachments/assets/c2884c4f-7ff8-426a-8bdd-89793d063d00)

**Comment by bohte** [comment]
Hi Armaghan,

As discussed I'm also adding description for bulk deletion API endpoint:

**Description**:
The idea of this endpoint would be to delete payments from invoices in a bulk manner
(_as well in a single call, as well in single transaction_)
Essentially, it will be used to kind of revert changes done via /post_payments endpoint introduced by you.

**Scenario**:
After user linked imported transaction to invoices (which resulted in payments creation under one or several invoices), in case it will be needed to perform unlinking - all the created payments must be deleted.
Under the hood, of course, all the accompanying rows (counter account row, fx adjustment row) must be removed as well.

**Tech design**:
For me it is difficult to say how it will be easier to handle on SQL-Ledger side, so i want to ask you to provide suggestions what request data u will need to execute deletion properly.
From our side, the most convinient way of handling it would be sending just invoice type (AR/AP) and imported transaction id which is set in the new column ("acc_trans" table) column introduced earlier.
The outcome of such request processing should be all the invoices, containing any linked payments, be updated and payments - removed.

**Reference**:
Relyable reference in this case would be ar/ap invoice edit form:
![image](https://github.com/user-attachments/assets/ae36bec6-9f13-44ce-bc61-cdd8b9fd702c)
Let's say that first payment row is linked to imported transaction, outcome of calling the endpoint described aboce should be the same as if I would clean the row, press update button and the post invoice without this row (and I would do it for several invoices at the same time)

**Comment by ledger123** [comment]
Hi @bohte ,

I have added the delete_payment route and also added the request format in the api.pl.  I have tested the code with the request examples provided.

Please test at your convenience and confirm all is ok.

Regards

**Comment by bohte** [comment]
Hi Armaghan,

I've retested everything and here are my findings (all the details are in our chat as agreed):
1) Payment posting does not work from "edit invoice" SQL-Ledger form
2) Invoice payments posting via new API endpoint resulting in missing decimals in
- "ar"/"ap" table 'paid' and 'fxpaid' columns
- "acc_trans" table counter account row 'amount' column (origin row 'amount' is ok)
3) Invoice payments deletion via new API endpoint does not change 'fxpaid' column value in "ar"/"ap" table

**Comment by ledger123** [comment]
Hi @bohte , Above issues are fixed.

Please test and confirm.

Regards

**Comment by bohte** [comment]
Hi Armaghan,

Here are my testing results:

1. Posting invoice changes via SL form works,
**BUT** imported_transaction_id values are not preserved for payment rows (values are erased after i edit invoice)

2. Problem with decimals when posting payments via new bulk API is still reproducible (moreover, i noticed that it doesn't just miss decimals, it also removes it even if it was there previously)

3. Setting 0 for fxpaid/paid columns is happening after deletion API is called
**BUT** now it is hardcoded and even if invoice had one more payment (not linked to transaction), fxpaid/paid is being set to 0

**Comment by bohte** [comment]
Testing results:
1) issue with posting invoice via SL form:
![image](https://github.com/user-attachments/assets/76355769-22e8-483c-9965-16795e003993)

2) when post invoice with payments without any changes - imported transaction id is only reapplied to original row, not to counter account row
 
3) issue with decimals is still reproducible

**Comment by ledger123** [comment]
Hi @bohte ,

All 3 issues are resolved and I have done testing on my server.

Please test in your convenience and let me know your results.

Regards
